### PR TITLE
arch-riscv: Implement Zcmp instructions

### PR DESCRIPTION
**Pull request by rogerchang23424**
1. Implement Zcmp(cm.push, cm.pop, cm.popret, cm.popretz, cm.mva01s, cm.mvsa01) instructions

2. The Zcd instructions overlap the Zcmp and Zcmt instruction. This option is used to enable/disable Zcd extension, implies enable Zcmp/Zcmt extension. If Zcd is enable, the Zcmp and Zcmt is disabled. Otherwise, Zcmp and Zcmt is enabled.

Spec: https://github.com/riscv/riscv-isa-manual/blob/main/src/zc.adoc

Change-Id: I3788eb6539e13a210c9946efc43ca1fef4639560

### CODE REVIEW

**Comment by rogerchang23424** [comment]
Hope the PR https://github.com/gem5/gem5/pull/1316 can be merged ASAP to avoid conflict

**Comment by powerjg** [comment]
#1316 has been merged

**Comment by rogerchang23424** [comment]
I will create another PR for Zcmt instructions

**Comment by rogerchang23424** [comment]
Tested these instructions internally

File: src/arch/riscv/insts/zcmp.cc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+CmMacroInst::CmMacroInst(
+    const char* mnem, ExtMachInst machInst, OpClass opClass)
+    : RiscvMacroInst(mnem, machInst, opClass), rlist(machInst.rlist)
+{
+}
+
+// Ref: https://github.com/riscv-software-src/riscv-isa-sim/blob/f7d0dba60/
+//      riscv/decode.h#L168
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wimplicit-fallthrough"
```
Comment (powerjg) [inline src/arch/riscv/insts/zcmp.cc]: Why not just add `[[fallthrough]];` I would prefer to not have these pragmas if at all possible as they are not portable across compilers/versions

File: src/arch/riscv/insts/zcmp.cc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+uint64_t
+CmMacroInst::stackAdj() const
+{
+  uint64_t stack_adj_base = 0;
+  switch (machInst.rlist) {
+  case 15:
+    stack_adj_base += 16;
+  case 14:
+    if (machInst.rv_type == RV64)
+      stack_adj_base += 16;
```
Comment (powerjg) [inline src/arch/riscv/insts/zcmp.cc]: Also, I would strong prefer `{ }` and the spacing is not 4 spaces.

**Review by powerjg — COMMENTED** [review]
Mostly seems OK. One small comment below. Thanks for the contribution!

**Review by BobbyRBruce — APPROVED** [review]

**Review by Harshil2107 — APPROVED** [review]
lgtm

**Comment by rogerchang23424** [comment]
Hi maintainers,

Is the PR ready to merge?

**Pull request merged**